src/main/python/macro_tab.py

```diff
@@ -122,6 +130,52 @@ def on_move(self, obj, offset):
         self.lines[other].insert(other)
         self.changed.emit()
 
+    def on_text_window(self):
+
+        # serialize all actions in this tab to a json
+        macro_text = []
+        macro_text.append([act.save() for act in self.actions()])
+        macro_text = json.dumps(macro_text[0])
+
+        textbox = TextboxWindow(macro_text, "vim", "Vial macro")
+
+        if textbox.exec():
+            macro_text = textbox.getText()
+            if len(macro_text) < 6:
+                macro_text = "[]"
+            macro_load = json.loads(macro_text)
+
+            # ensure a list exists
+            if not isinstance(macro_load, list):
+                return
+
+            # associate action types with tags from the macro json
+            tag_to_action = {
+                "down": ActionDown,
+                "up": ActionUp,
+                "tap": ActionTap,
+                "text": ActionText,
+                "delay": ActionDelay,
+            }
+            tag_to_actionUI = {
+                "down": ActionDownUI,
+                "up": ActionUpUI,
+                "tap": ActionTapUI,
+                "text": ActionTextUI,
+                "delay": ActionDelayUI,
+            }
```

Comment:
Can you reduce the duplication here, there's already a similar implementation in `restore_macros` which restores from json into a macro (that specific part could be moved into a separate utility function); and then a mapping of action to UI element is available in macro_recorder.py->ui_action.

Still not ideal and macro code overall is messy, but what I want to ensure is that when adding another macro type that saving/restoring functionality doesn't have to be duplicated in yet another place.

Reply:
That makes sense, I will work on reducing the duplication. KISS

Reply:
https://en.wikipedia.org/wiki/KISS_principle

Reply:
Commit baf603a addresses this change